arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/CircuitBreaker.kt

```diff
@@ -40,10 +41,36 @@ class CircuitBreaker constructor(
       is State.HalfOpen -> curr.awaitClose.get()
     }
 
-  /** Returns a new task that upon execution will execute the given
+  /**
+   * Returns a new task that upon execution will execute the given
+   * task, but with the protection of this circuit breaker.
+   */
+  @Deprecated("#protect is being renamed to protectOrThrow", ReplaceWith("protectOrThrow(fa)"))
+  suspend fun <A> protect(fa: suspend () -> A): A =
+    protectOrThrow(fa)
+
+  suspend fun <A> protectEither(fa: suspend () -> A): Either<Throwable, A> =
+    Either.catch { protectOrThrow(fa) }
+
+  /**
+   * Returns a new task that upon execution will execute the given
    * task, but with the protection of this circuit breaker.
+   * If an exception in [fa] occurs it will be rethrown
    */
-  tailrec suspend fun <A> protect(fa: suspend () -> A): A =
+  suspend fun <A> protectOrNull(fa: suspend () -> A): A? =
+    try {
+      protectOrThrow(fa)
+    } catch (ex: Throwable) {
```

Comment:
```suggestion
    } catch (ex: ExecutionRejected) {
```

I'm wondering for `protectOrNull`, it should return `null` when `fa` was rejected or it should throw the original exceptions. For example for `fa = suspend { throw RuntimeException() }` we expect `RuntimeException` and not `null`.

cc\\ @raulraja @franciscodr

Reply:
I think original exceptions should be thrown as `null` does not contain enough information as to what happened there and the `null` return can be a normal return of a user computation that returns null. Returning null in all cases won't allow to disambiguate whether a task finished exceptionally or not and we don't have here Either or anything similar to get a proper type info about exceptional cases.